hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java

```diff
@@ -962,6 +963,11 @@ public static boolean scmInit(OzoneConfiguration conf,
         return false;
       }
     } else {
+      // If SCM HA was not being used before pre-finalize, and is being used
+      // when the cluster is pre-finalized for the SCM HA feature, init
+      // should fail.
+      ScmHAUnfinalizedStateValidationAction.checkScmHA(conf, scmStorageConfig);
+
       clusterId = scmStorageConfig.getClusterID();
       final boolean isSCMHAEnabled = scmStorageConfig.isSCMHAEnabled();
       if (SCMHAUtils.isSCMHAEnabled(conf) && !isSCMHAEnabled) {
```

Comment:
Hi @errose28 , I noticed that this check is just the same as `checkScmHA()`, so this block of code is somewhat 'dead' after `checkScmHA()` ?

Reply:
If SCM HA is used pre-finalized and it was not already being used, this action will fail with an exception and the code after the `checkScmHA` call will not run because the whole SCM will exit. However, if SCM HA is used pre-finalized when it was being used before the upgrade framework, or the cluster is not pre-finalized, the action will not throw an exception, and the code below the call will run. So it is not totally dead. This is the same behavior added to the SCM start method. Raising the exception on failure is preferable to an if-else check because the `UpgradeException` with `PREFINALIZE_ACTION_VALIDATION_FAILED` result code and message allows us to give more information to the user, rather than returning `false` from init.